### PR TITLE
Customize page: show calendar date if present

### PR DESCRIPTION
**Pull request by leccelecce**
Take advantage of the yaml format to display the calendar date (in Fri 01 Jul 2023 format) on the lineup customization page if present.

I've done various testing around different edge cases and it should work fine alongside existing lineups.

This is a very naive implementation, and it will log a warning if the number of calendar dates doesn't equal the number of days and fall back to the simple day number.

If you could re-run the cache warm script for Nos Alive 2023 that should prove it's working correctly.

### CODE REVIEW

**Comment by leccelecce** [comment]
Ran formatter on changed *.ts files - now ready

**Review by dot4qu — CHANGES_REQUESTED** [review]
Okay, overall super great and thanks for finally taking the step to actually use the first-class yaml data because I've been too lazy.

Instead of taking this intermediate date where you're pulling out the date field from the Day yaml object but still just saving and passing around a stringified list of raw strings or numbers, I'd prefer to implement the actual use of the full Day metadata.

You can start with just adding support for using the date field from the object and we can expand the functionality to store and use stuff like display_name, etc, in the future, if you want since that would keep the scope the same. But if you implement it with the actual object storage and retrieval now it'll make using the rest of the fields later super easy to add. 

I was all over the place jumping around explaining different parts in the comments so lemme know if what I'm suggesting is confusing.

**Comment by dot4qu** [comment]
Also since that's a pretty decent amount more work I'm happy to merge this to add date support for now, and then follow it up with another PR to support the full metadata format. My only hesitation with merging this as-is is that it would add this `:dates` key which would have to be cleaned up in the future. And if you switch it to `:days` like I'd like to have it, that would require backwards compat logic to support either parsing this list of dates or the actual objects, depending on what's stored, in the future

File: views/customize-list.handlebars [hunk 1]
```diff
@@ -48,7 +48,12 @@
                             <div>
                                 <label for="{{this.obj}}" class="pure-checkbox">
                                 <input class="dayCheckbox" id="{{this.obj}}" type="checkbox" value="{{this.obj}}" onchange="refreshArtistList()" {{state}}>
-                                {{this.obj}}</label>
+                                {{#if ../dates.length}}
+                                    {{formatDate (lookup ../dates @index)}}
```
[the remark this answers is not in the document]
Comment (leccelecce) [reply views/customize-list.handlebars:52]: > This is kind of annoying and I apologise for it but I hate how this displays dates by default haha. Is there a way to pass a format string to it to make it look nicer? Or do we have to have our own formatDate handler registered in the code to do that

How about "Thursday July 6th" instead of "Thu Jul 06 2023"? Year can be dropped because it's already at the top of the page.

File: src/redis-helper.ts [hunk 2]
```diff
@@ -5,6 +5,31 @@ import * as mbHelper        from "./mb-helper";
 import * as setlistFmHelper from "./setlist-fm-helper";
 import * as spotifyHelper   from "./spotify-helper";
 
+export async function getLineupDates(redisClient: redis.RedisClient, festivalName: string, festivalYear: number):
+    Promise<Date[]> {
+    const datesPromise: Promise<string[]> = new Promise((resolve, reject) => {
+        redisClient.get(`festival:${festivalName.toLowerCase()}_${festivalYear}:dates`, (err: Error, obj: string) => {
```
[the remark this answers is not in the document]
Comment (leccelecce) [reply src/redis-helper.ts:11]: Re 2) and display name format - I had this in for flexibility for festivals that do "Weekend 1 Sat", "Weekend 1 Sun" etc as an optional extra bit of information - not sure how useful it is so would probably not bother for the majority and only show it if present.

File: src/routes/pages.ts [hunk 3]
```diff
@@ -189,13 +189,22 @@ function setRoutes(redisClient: redis.RedisClient): express.Router {
...
+                festival.name} ${queryYear}`);
+            dates = [];
+        }
+
         res.render("customize-list", {
             prod : process.env.DEPLOY_STAGE === 'PROD',
             titleOverride : `Customize Playlist - ${festival.display_name} ${queryYear}`,
             festival,
             festivalYear : queryYear,
+            dates,
```
[the remark this answers is not in the document]
Comment (leccelecce) [reply src/routes/pages.ts:207]: I started looking at this and it's a bit fiddly, because currently `daysMap` is actually built up through iterating through the artist list, and the days are essentially extracted from there, whereas now we would be going to Redis to get the days and using that as the canonical list. Any thoughts on how you want that to work? I'm thinking probably get the days+metadata from redis before the `for (const artist of artists)` loop, but then need to look at all the checkbox logic too.

**Comment by leccelecce** [comment]
> Also since that's a pretty decent amount more work I'm happy to merge this to add date support for now, and then follow it up with another PR to support the full metadata format. Also since this is all going into dev we can merge in piecemeal support (i.e. redis support first, then display support in another PR), it doesn't have to actually work all from one PR.
> 
> My only hesitation with merging this as-is is that it would add this `:dates` key which would have to be cleaned up in the future. And if you switch it to `:days` like I'd like to have it, that would require backwards compat logic to support either parsing this list of dates or the actual objects, depending on what's stored, in the future

I've pretty much worked up all the changes proposed above so happy to just round them off and do it properly :-)

File: src/routes/pages.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (leccelecce) [reply src/routes/pages.ts:207]: Nevermind the above, will push some proposed changes.

**Comment by leccelecce** [comment]
Made a miss of this in git whilst rebasing - will fix it up then mark as ready...

File: src/warm-cache-for-festival.ts [hunk 4]
```diff
@@ -133,8 +125,25 @@ async function warm(festival: string, years: number[]) {
...
+            const allDates: Date[] = data.days.map(day => day.date);
+
             // load the days/lineups, per the text format
             for (const day of data.days) {
+                day.date = new Date(day.date);
+
+                // Save day metadata
+                redisClient.hmset(`festival:${festival.toLowerCase()}_${year}:days:${day.number}`,
+                                  {
+                                      number : day.number.toString(),
```
Comment (dot4qu) [inline src/warm-cache-for-festival.ts:137]: This probably doesn't need to be stringified but I'm ambivalent since it's redundant already because it's in the key needed to even query for this object.

File: src/warm-cache-for-festival.ts [hunk 5]
```diff
@@ -143,12 +152,10 @@ async function warm(festival: string, years: number[]) {
...
                 }
             }
-        }
 
-        // Save all days so we can pull them later
-        redisClient.set(`festival:${festival.toLowerCase()}_${year}:days`,
-                        JSON.stringify(Object.keys(artistObjs)),
-                        redis.print);
+            // Save all days so we can pull them later
+            redisClient.set(`festival:${festival.toLowerCase()}_${year}:days`, JSON.stringify(allDates), redis.print);
```
Comment (dot4qu) [inline src/warm-cache-for-festival.ts:157]: I don't think we can have two different types of objects saved for this key. 
Above in the .txt case it's saving a stringified list of the day numbers, i.e. "[1,2,3]" (same as the previous behavior). Here in the yaml case, it's saving a stringified list of the actual date string supplied from the yaml file, i.e. "["2022-01-01", "2022-01-02", ...]". 

I'd like to stick with just the day number representing days as much as possible, since it's like lowest common denominator. So the `allDates` list isn't necessary here, and day metadata (if available) will be accessed with the day number in the key, like you have it above already. You agree?

File: views/customize-list.handlebars [hunk 6]
```diff
@@ -46,9 +46,14 @@
...
-                                <label for="{{this.obj}}" class="pure-checkbox">
-                                <input class="dayCheckbox" id="{{this.obj}}" type="checkbox" value="{{this.obj}}" onchange="refreshArtistList()" {{state}}>
-                                {{this.obj}}</label>
+                                <label for="{{this.obj.number}}" class="pure-checkbox">
+                                <input class="dayCheckbox" id="{{this.obj.number}}" type="checkbox" value="{{this.obj.number}}" onchange="refreshArtistList()" {{state}}>
+                                {{#if this.obj.date}}
+                                    {{this.obj.date}}
+                                {{else}}
+                                    {{this.obj.number}}
+                                {{/if}}
```
Comment (dot4qu) [inline views/customize-list.handlebars:55]: 👍🏼

File: src/routes/pages.ts [hunk 7]
```diff
@@ -145,6 +145,24 @@ function setRoutes(redisClient: redis.RedisClient): express.Router {
         const mainGenresMap: Map<string, StatefulObject>     = new Map<string, StatefulObject>();
         const specificGenresMap: Map<string, StatefulObject> = new Map<string, StatefulObject>();
         const daysMap: Map<string, StatefulObject>           = new Map<string, StatefulObject>();
+
+        const dayNumbers: number[] = await redisHelper.getLineupDays(redisClient, festival.name, queryYear);
+
+        const daysWithMetadata: LineupDay[] = await Promise.all(dayNumbers.map(
+            async (day) => redisHelper.getLineupDayMetadata(redisClient, festival.name, queryYear, day)));
+
+        for (const lineupDay of daysWithMetadata) {
```
Comment (dot4qu) [inline src/routes/pages.ts:154]: I'm not totally sure if `daysWithMetadata` can be null or if it will always be a list returned from Promise.all, but even if it's a list it's possible to be a list of nulls, which will error as soon as `lineupDay` is accessed. The two redis functions `getLineupDays` and `getLineupDayMetadata` should either be changed to always return an empty list if the final if-check fails in the function (instead of just not returning anything) or everything that calls those functions needs to be null-checked, so checking `dayNumbers` before dayNumbers.map is called and somehow checking the contents of what's returned by Promise.all

**Comment by leccelecce** [comment]
Thanks for the review - haven't actually finished changes hence putting the PR back into draft, sorry! I need to do some more testing locally on handling edge cases then re-push and mark PR as ready
